commit 9935e1390421f33b409e7b27ad43c3201aa3cbb0 (refs/changes/65/4465/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-06-29T17:12:01-07:00 (1 year, 3 months ago)
    
    MANTA-3804 waferlock should retry after getting no NICs on a SAPI VM

diff --git a/lib/sapi.js b/lib/sapi.js
index 967579e..cce5e31 100644
--- a/lib/sapi.js
+++ b/lib/sapi.js
@@ -60,6 +60,7 @@ function SapiPoller(options) {
 	this.sp_errdelay = options.minPoll;
 
 	this.sp_insts = {};
+	this.sp_instRetries = {};
 	this.sp_lastError = undefined;
 
 	this.sp_agent = options.agent;
@@ -257,13 +258,26 @@ SapiPoller.prototype.state_runq_do = function (S) {
 
 		mod_assert.arrayOfObject(objs);
 
+		var retryEnt = { dc: ent.dc, vms: [] };
+
 		objs.forEach(function (vm) {
 			if (vm.state === 'destroyed' || vm.destroyed)
 				return;
 			if (vm.state === 'failed')
 				return;
-			if (!vm.nics || !Array.isArray(vm.nics))
+			if (!vm.nics || !Array.isArray(vm.nics) ||
+			    vm.nics.length < 1) {
+				self.sp_log.warn({ uuid: vm.uuid }, 'got a ' +
+				    'vmapi VM with no valid NICs');
+				if (self.sp_instRetries[vm.uuid] === undefined)
+					self.sp_instRetries[vm.uuid] = 0;
+				var retries = self.sp_instRetries[vm.uuid];
+				if (retries < 5) {
+					++self.sp_instRetries[vm.uuid];
+					retryEnt.vms.push(vm.uuid);
+				}
 				return;
+			}
 
 			var ips = [];
 			vm.nics.forEach(function (nic) {
@@ -281,6 +295,14 @@ SapiPoller.prototype.state_runq_do = function (S) {
 			self.sp_pool.refreshTag('sapi:' + vm.uuid, ips);
 		});
 
+		if (retryEnt.vms.length > 0) {
+			self.sp_vmq.push(retryEnt);
+			self.sp_lastError = new VError('VMAPI data was ' +
+			    'missing NIC information, retrying some VMs');
+			S.gotoState('error');
+			return;
+		}
+
 		S.gotoState('runq');
 	}));
 };
